PagingWithNetworkSample/app/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/repository/inMemory/byItem/ItemKeyedSubredditPagingSource.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byItem
+
+import androidx.paging.LoadType.*
+import androidx.paging.PagingSource
+import androidx.paging.PagingSource.LoadResult.Page
+import androidx.paging.PagingState
+import com.android.example.paging.pagingwithnetwork.reddit.api.RedditApi
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import java.io.IOException
+
+/**
+ * A [PagingSource] that uses the "name" field of posts as the key for next/prev pages.
+ *
+ * Note that this is not the correct consumption of the Reddit API but rather shown here as an
+ * alternative implementation which might be more suitable for your backend.
+ *
+ * @see [com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byPage.PageKeyedSubredditPagingSource]
+ */
+class ItemKeyedSubredditPagingSource(
+        private val redditApi: RedditApi,
+        private val subredditName: String,
+        private val networkDispatcher: CoroutineDispatcher
+) : PagingSource<String, RedditPost>() {
+
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, RedditPost> {
+        return when (params.loadType) {
+            REFRESH -> loadInitial(params)
+            // Ignored, since we only ever append to our initial load.
+            START -> throw IllegalStateException("Ignored")
+            END -> loadAfter(params)
+        }
+    }
+
+    private suspend fun loadAfter(params: LoadParams<String>): LoadResult<String, RedditPost> {
```

Comment:
unify these, similar to PageKeyed, should just need the one load method

Reply:
They call different RedditAPI methods though - I think having helpers is justified? Let me reduce some copied code though.

Reply:
Nevermind - just saw the optional params suggestion. Done.